tools/autograd/templates/VariableType.cpp

```diff
@@ -311,30 +302,33 @@ static void rebase_history(Variable& var, std::shared_ptr<Function> grad_fn, int
 
 // var must be the only differentiable output of the function. Use the ArrayRef
 // overload for functions with multiple differentiable outputs.
-static void set_history(Variable& var, std::shared_ptr<Function> grad_fn, int output_nr=0) {
+static void set_history(Tensor& t, std::shared_ptr<Function> grad_fn, int output_nr=0) {
+  auto& var = static_cast<Variable&>(t);
   if (grad_fn) {
     grad_fn->num_inputs = 1;
     var.get()->output_nr = output_nr;
     var.get()->_grad_fn = std::move(grad_fn);
   }
 }
 
-static void set_history(at::ArrayRef<Variable> vl, std::shared_ptr<Function> grad_fn) {
+static void set_history(at::ArrayRef<Tensor> tl, std::shared_ptr<Function> grad_fn) {
   if (grad_fn) {
-    grad_fn->num_inputs = vl.size();
+    grad_fn->num_inputs = tl.size();
     int64_t output_nr = 0;
-    for (auto& var : vl) {
-      if (!var.defined()) continue;
+    for (auto& t : tl) {
+      if (!t.defined()) continue;
       // TODO: combine this with the Variable construction
+      auto& var = static_cast<const Variable&>(t);
       var.get()->output_nr = output_nr;
       var.get()->_grad_fn = grad_fn;
       output_nr++;
     }
   }
 }
 
-static variable_list flatten(const TensorList& tensors) {
-  return cast_tensor_list(tensors);
+static at::ArrayRef<Variable> flatten(TensorList tensors) {
```

Comment:
This micro-optimization makes me nervous, as none of the overloads return (nor can they return) an ArrayRef...

Reply:
It's not even valid. `tensors` will be a temporary copy of an argument, so the data of this array ref will be freed when returning.

Reply:
Oh yeah, I remember; the output is live only as long as the input is live, and that's a nontrivial invariant to remember especially since it would only apply to this argument.